test/test_bindings_module.py

```diff
@@ -409,3 +409,107 @@ def test_get_mesh_vertices_and_ids(self):
         fake_ids, fake_coordinates = solver_interface.get_mesh_vertices_and_ids(fake_mesh_id)
         self.assertTrue(np.array_equal(fake_ids, vertex_ids))
         self.assertTrue(np.array_equal(fake_coordinates, coordinates))
+
+    def test_is_gradient_data_required(self):
+        solver_interface = precice.Interface("test", "dummy.xml", 0, 1)
+        fake_bool = 0  # compare to output in test/SolverInterface.cpp
+        fake_data_id = 0
+        self.assertEqual(fake_bool, solver_interface.is_gradient_data_required(fake_data_id))
+
+    def test_write_block_scalar_gradient_data(self):
+        solver_interface = precice.Interface("test", "dummy.xml", 0, 1)
+        write_data = np.array([[1, 2, 3], [6, 7, 8], [9, 10, 11]])
+        solver_interface.write_block_scalar_gradient_data(1, np.array([1, 2, 3]), write_data)
```

Comment:
This test is not really testing the bindings API, and I understand that the problem is because gradient data functions are only write functions and we dont have read functions to get data back and check if the correct data was passed on to the preCICE API. Lets do the following, write the `write_data` to the mocked preCICE interface and then access the `fake_gradient_buffer` inside the mocked interface to check if the data is equal to `write_data`. The check can be done with `self.assertTrue(np.array_equal(write_data, solver_interface.fake_gradient_buffer))`.

**Note: This applies to all the tests below**

Reply:
It seems we need another solution since on the cpp side `fake_gradient_buffer` is just a static variable 
```    self.assertTrue(np.array_equal(write_data, solver_interface.fake_gradient_buffer))
AttributeError: 'cyprecice.Interface' object has no attribute 'fake_gradient_buffer'

```

Reply:
Okay I understand the problem. Thinking more on this I have concluded that the only way to test such write-only functionality is to add something to the Python bindings API itself. Lets add a private function to the Python bindings API which returns the gradient data written. This function will be used only for testing and will be hidden from the user. @kursatyurt can you find a nice way to do this?

Reply:
Free function sounds like a good idea. Which errors do you face? Can you post them here? If you define (and use) the free function only in the `cpp` file I would consider wrapping it into an anonymous namespace.

Reply:
The idea of having a free function in the local header file sounds really good! That way we can use it for testing and do not have to add anything in the bindings API. Lets debug the errors you are facing :+1:

Reply:
I have a simpler idea to handle the problem we are facing. Why dont we just write the gradient data to the `fake_read_write_buffer` and then return is using the read commands? That way we can check if the gradient data is correct inside the API and dont have to implement additional read functionality anywhere.

Reply:
@IshaanDesai block versions would work, other way around it gets a bit more complex, I will change to this. Thank you very much

Reply:
> block versions would work, other way around it gets a bit more complex

What do you mean by the *other way around* here? I would just use the corresponding read functionality, so if you are testing `write_block_scalar_gradient_data`, I would use `read_block_scalar_data` to get the data back, and same for all other other API functions.

Reply:
> > block versions would work, other way around it gets a bit more complex
> 
> What do you mean by the _other way around_ here? I would just use the corresponding read functionality, so if you are testing `write_block_scalar_gradient_data`, I would use `read_block_scalar_data` to get the data back, and same for all other API functions.

Writing new functions to return gradient data and keep them updated. 

I have used `read_block_scalar_data` for all of them. The problem here is dimensions of return arrays do not match since the gradient of the scalar is a vector, etc. 

It is a bit ugly, but I have to convert all the input to a 1D nd.array for comparison.

Now mostly works. I have problem with `test_write_block_scalar_gradient_data_single_float`. I ended up with a segmentation fault. Could you check that?

Reply:
> Now mostly works. I have problem with `test_write_block_scalar_gradient_data_single_float`. I ended up with a segmentation fault. Could you check that?

Fixed :+1:  I'll clean up the tests a bit before merging this